### ISSUE TITLE
Esi support

### ISSUE THREAD

**Issue by eddycharly** [auxmoney/OpentracingBundle-core#21]
Do you have an idea of what should be done to support esi requests ?
Currently they are tracked as separate traces, not related to the parent trace. 
Thanks

**Comment by cawolf** [comment]
Hi @eddycharly,
we took some time this day to investigate the issue and can confirm, that direct requests to a Symfony instance will result in 2 or more unrelated traces (template trace + all fragment traces).

Upon debugging the [Symfony implementation](https://github.com/symfony/symfony/blob/v3.4.38/src/Symfony/Component/HttpKernel/HttpCache/Esi.php#L48-L61) and catching up on the [ESI language specification](https://www.w3.org/TR/2001/NOTE-esi-lang-20010804), there will unfortunately be no standardized way to implement a solution in the OpentracingBundle, as the language specification has no concept of passing headers from the template request to the fragments.

However, the template request can pass query parameters to fragments. But this is a non-standard way of propagating headers and will not be standardized by the [upcoming W3C specification](https://w3c.github.io/trace-context/#abstract), which is why we will not consider query propagation as a workaround for this issue.

Fortunately there is a possible way to prevent the observed disconnect between template and fragment traces, and it is described in the ESI language specification (under "6. Protocol Considerations"):
> When an ESI template is processed, a separate request will need to be made for each include encountered. Implementations may use the original request's headers (e.g., Cookie, User-Agent, etc.) when doing so. Additionally, response headers from fragments (e.g., Set-Cookie, Server, Cache-Control, Last-Modified) may be ignored, and should not influence the assembled page.

This means that if you have any kind of trace-generating layer before hitting an ESI enabled page, you *may* be fine, because the span context *may* be propagated by headers to both the template and the fragments. We could reproduce that behaviour by using an instrumented command, which performed a Guzzle request to our initial test project with ESI, and generated this trace:

![image](https://user-images.githubusercontent.com/1932623/76021043-d5e16000-5f24-11ea-92bf-0669a0ce0958.png)

But as stated, it depends heavily on the ESI implementation, meaning it can work with the Symfony built-in application, but won't with Squid or Varnish (we did not test it today). You will need to find out if your used technology is capable of passing the original headers to all fragments.

So, in conclusion: if the trace is generated before (by an instrumented request, command or other applications like [Traefik](https://docs.traefik.io/observability/tracing/overview/) or [Nginx](https://medium.com/opentracing/how-to-enable-nginx-for-distributed-tracing-9479df18b22c)) and passed to an ESI page, chances are good that you will get connected traces. If not, our bundle has no power to fix that. Is this answering your question?

Best regards,
@cawolf

**Comment by eddycharly** [comment]
Hi @cawolf
This is definitely answering my question, thanks for the very detailed explanation.

We came to the same conclusion (with less details and analysis though) yesterday after enabling opentracing on the nginx server in front of the symfony app.
All the traces were correctly connected to the parent trace started by the nginx server.

It would be nice to keep track of this information around in docs. I will try to take time for this and open a PR.

Thank you very much.

**Comment by cawolf** [comment]
Great! I also thought about documenting it, and I'm looking forward to your PR. :)

**Comment by eddycharly** [comment]
Done.
I tried to organize things a little bit, don't hesitate if it's not clear or not well structured.

**Comment by cawolf** [comment]
LGTM; thank you for your contribution!
